src/python/pants/base/exception_sink.py

```diff
@@ -43,11 +44,38 @@ def signal_handler_mapping(self):
     # instead just iterating over the registered signals to set handlers, so a dict is probably
     # better.
     return {
-      signal.SIGINT: self.handle_sigint,
+      signal.SIGINT: self._handle_sigint_if_enabled,
       signal.SIGQUIT: self.handle_sigquit,
       signal.SIGTERM: self.handle_sigterm,
     }
 
+  def __init__(self):
+    self._ignore_sigint_lock = threading.Lock()
+    self._ignore_sigint = 0
+
+  def _check_sigint_gate_is_correct(self):
+    assert self._ignore_sigint >= 0, \
+      "This should never happen, someone must have modified the counter outside of SignalHandler."
+
+  def _handle_sigint_if_enabled(self, signum, _frame):
+    with self._ignore_sigint_lock:
+      self._check_sigint_gate_is_correct()
+      should_ignore_sigint = self._ignore_sigint
```

Comment:
Both of the names on this line are kind of unclear...

`should_ignore_sigint` reads to me like a boolean (when actually the only "true" value of it is `0`, which is False as a boolean)
`self._ignore_sigint` also reads to me like a boolean.

Maybe we could call both of these `number_of_threads_ignoring_sigint` or similar?